### PR TITLE
Fix OpenGL segmentation fault on Linux

### LINKED ISSUE #36220: Mantid Crashes when trying to use Instrument view OpenGL 3D view on Rocky IDAaaS
**Describe the bug**
Mantid Crashes when trying to use Instrument view OpenGL 3D view on Rocky IDAAS.

**To Reproduce**
1) Create an IDAaaS Rocky 8 Instance with **no GPU**: https://dev.analysis.stfc.ac.uk/
2) Load some data (I Used MAR11060).
3) Open Instrument View
4) Under Display Settings, select `Use OpenGL`.
5) Using the top drop down, change from `Cylindrical Y` to `Full 3D`.
6) Observe crash.

**Expected behavior**
Mantid should not crash.

**Platform/Version (please complete the following information):**
IDAaaS Rocky8

### PR DESCRIPTION
**Pull request by jclarkeSTFC**
When drawing some instruments in full 3d (with OpenGL), putting all the draw commands for the instrument in one big list seems to cause a segmentation fault on certain versions of Linux, including the version used on IDAaaS. This changes `InstrumentRenderer` to draw each component of the instrument using a separate list, when the `MantidOptions.InstrumentView.MesaBugWorkaround` option in the user properties file is set to `On`.

The original bug in Mesa was fixed [here](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/27556), but that version has not made it onto IDAaaS yet.

#### Summary of work

- Draw each instrument component in a separate display list when the new option is enabled.
- If option is disabled then behaviour should be the same.
- Add a method that will give you the OpenGL version string (someone will thank me in the future).
- Split the picking and non-picking display lists into named member variables instead of using two indices.

Fixes #36220 

### To test:
See #36220, or what caused the original segmentation fault for me was to view full 3D for MAR11001 (on Rocky 8 IDAaaS and Ubuntu). Other instruments and files will need checking as well.

Updated conda packages have bee uploaded to thomashampson/label/mesa_workaround:
`mamba install -c thomashampson/label/opengl_segfault_fix mantidworkbench`
Install this on IDAaaS and check that it doesn't crash.

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Review by jhaigh0 — CHANGES_REQUESTED** [review]
One small thing.
I find it not so easy to ready, but I reckon that's mostly just because its complex in nature. I wonder if some comments explaining how it works would be helpful? I'll let other reviewers chime in, I don't mind either way.

**Comment by RichardWaiteSTFC** [comment]
Tested this on MARI > Excitations Single Crystal (no GPU) and found no crash (started from terminal with `mantidworkbench`)

File: qt/widgets/instrumentview/src/InstrumentRendererClassic.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  OpenGLError::check("InstrumentActor::draw()");
+}
+
+void InstrumentRendererClassic::draw(const std::vector<bool> &visibleComps, bool showGuides, bool picking) {
+  const auto &compInfo = instrActor().componentInfo();
+  std::vector<bool> visited(compInfo.size(), false);
+
+  for (size_t i = compInfo.root(); i != std::numeric_limits<size_t>::max(); --i) {
+    drawComponent(i, visibleComps, showGuides, picking, compInfo, visited);
+  }
```
Comment (jhaigh0) [inline qt/widgets/instrumentview/src/InstrumentRendererClassic.cpp:62]: I know this has been copied from the old code so don't worry with fixing it now, but there's got to be a better way to do this 🙃 I think the component info class should have a way of retrieving the max index

File: qt/widgets/instrumentview/src/InstrumentRendererMultiList.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+  m_useNonPickingDisplayList = false;
+  m_usePickingDisplayList = false;
+
+  m_pickingDisplayListId.resize(componentInfo.size());
+  m_nonPickingDisplayListId.resize(componentInfo.size());
+  for (size_t i = 0; i < componentInfo.size(); ++i) {
+    m_nonPickingDisplayListId[i] = 0;
+    m_pickingDisplayListId[i] = 0;
+  }
```
Comment (jhaigh0) [inline qt/widgets/instrumentview/src/InstrumentRendererMultiList.cpp:27]: There might be a simpler way to initialise these vectors
```
std::vector<T> myVector(length, repeatingElementOfTypeT);
```

File: qt/widgets/instrumentview/src/InstrumentRendererMultiList.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    if (i != 0) {
+      glDeleteLists(i, 1);
+    }
+  }
+}
+
+void InstrumentRendererMultiList::renderInstrument(const std::vector<bool> &visibleComps, bool showGuides,
+                                                   bool picking) {
+  if (std::none_of(visibleComps.cbegin(), visibleComps.cend(), [](bool visible) { return visible; }))
+    return;
```
Comment (jhaigh0) [inline qt/widgets/instrumentview/src/InstrumentRendererMultiList.cpp:46]:
```suggestion
  if (std::none_of(visibleComps.cbegin(), visibleComps.cend()))
    return;
```

File: qt/widgets/instrumentview/src/InstrumentRendererMultiList.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+  if (!instrActor().isInitialized()) {
+    return;
+  }
+
+  OpenGLError::check("InstrumentActor::draw()");
+  const auto &displayListIds = picking ? m_pickingDisplayListId : m_nonPickingDisplayListId;
+  auto &useDisplayList = picking ? m_usePickingDisplayList : m_useNonPickingDisplayList;
```
Comment (jhaigh0) [inline qt/widgets/instrumentview/src/InstrumentRendererMultiList.cpp:54]: Feels as if maybe these bools could be simplified? is it always either `m_usePickingDisplayList` or `m_useNonPickingDisplayList`? In that case could they be one bool? I might be misunderstanding.

**Review by jhaigh0 — CHANGES_REQUESTED** [review]
A few minor suggestions, mostly to do with the older code, but looking pretty much there

**Review by jhaigh0 — APPROVED** [review]
We've had a test in IDAaaS now that they've added the config option to the local properties. Prevents the crash as intended.

**Pull request merged**